File: src/simcore/conservation.py
"""Conservation ledger + the every-step balance gate (decision #13, step-alg #7).

The augmented system (modeled stocks + boundary reservoirs) is **closed**: every
Phase-0 state change is balanced — a flow has ``Σ legs == 0`` per quantity
(arbitration scales *whole* flows, preserving that), and extinction's loss-sink
routing is a balanced *non-flow* state change. So over the augmented system, per
asserted quantity the **total mass across all stocks is unchanged step-to-step**,
within tolerance::

    residual_q = Σ_{stocks s of quantity q} (after[s].amount − before[s].amount) ≈ 0

This is the **bug-resistant form** of the plan's ``inputs − outputs − ΔStored = 0``:
they are equal up to sign (``inputs − outputs = −boundary_delta``, so
``inputs − outputs − ΔStored = −(boundary_delta + stored_delta) = −residual``), but
the total-mass residual has **no sign/classification surface in the pass/fail** —
the ``StockKind`` partition feeds only the diagnostic decomposition.

This module reasons about **state deltas, not flows**. It deliberately does *not*
reuse ``flow.assert_flow_balanced`` over a synthetic "whole-step delta" flow, which
would mis-state the step-3 caveat that extinction's loss-sink routing is a balanced
*non-flow* change. It shares with ``simcore.flow`` only the contract constants
(``ASSERTED_QUANTITIES``, ``BALANCE_ATOL``/``BALANCE_RTOL``, ``ConservationError``) —
which is the real drift risk, and is covered.

Pure stdlib only.
"""

from dataclasses import dataclass

from simcore.flow import ConservationError
from simcore.quantities import (
    ASSERTED_QUANTITIES,
    BALANCE_ATOL,
    BALANCE_RTOL,
    Quantity,
    StockKind,
)
from simcore.state import State


@dataclass(frozen=True)
class QuantityLedger:
    """Per-quantity conservation accounting for one step (decision #13).

    - ``boundary_delta`` — ``Σ Δamount`` over BOUNDARY stocks of the quantity. This
      *is* the ledger's net Input/Output; the net flux *into* the modeled system is
      ``−boundary_delta``.
    - ``stored_delta`` — ``Σ Δamount`` over POOL + POPULATION stocks (the modeled
      ΔStored).
    - ``residual`` — ``boundary_delta + stored_delta`` (≡ total-mass Δ); ~0 when the
      step conserved the quantity. The decomposition is exact by construction (the
      residual is defined as the sum of the two partition deltas, not a separate
      interleaved sum), so ``boundary_delta + stored_delta == residual`` always holds.

    Clean accounting only — no ``scale`` field; the relative-tolerance basis is a
    tolerance detail computed in ``assert_conserved``.
    """

    quantity: Quantity
    boundary_delta: float
    stored_delta: float
    residual: float


def compute_ledger(before: State, after: State) -> tuple[QuantityLedger, ...]:
    """Per-quantity ledger for the step ``before → after`` (canonical quantity order).

    Covers every quantity *present* in the stocks, ``ENERGY`` included as a
    diagnostic (balance-exempt, decision #8 — mirroring ``flow.per_quantity_residual``).
    Per-stock deltas are accumulated within each ``StockKind`` partition in **sorted
    stock-id order** (decision #15: "every reduction"; float sums are non-associative,
    so this — not mapping iteration order — is what makes the ledger bit-identical
    under registration shuffle). The residual is the sum of the two partition deltas,
    so the decomposition is exact.

    Raises ``ValueError`` if ``before``/``after`` do not share the same stock-id key
    set: Phase 0 never adds or removes stocks mid-run, so a mismatch is an engine bug
    (``ValueError``, not ``assert`` — the latter is stripped under ``python -O``).
    """
    if before.stocks.keys() != after.stocks.keys():
        raise ValueError(
            "conservation ledger requires before/after to share the same stock ids; "
            "Phase 0 never adds/removes stocks mid-run "
            f"(before-only={set(before.stocks) - set(after.stocks)!r}, "
            f"after-only={set(after.stocks) - set(before.stocks)!r})"
        )
    boundary: dict[Quantity, float] = {}
    stored: dict[Quantity, float] = {}
    for sid in sorted(before.stocks):
        b = before.stocks[sid]
        delta = after.stocks[sid].amount - b.amount
        bucket = boundary if b.kind is StockKind.BOUNDARY else stored
        # Element-composition fold (P2.1) — the state-delta mirror of the leg fold
        # in ``flow.per_quantity_residual``. This module deliberately does NOT
        # reuse the leg path (it reasons over deltas, see the docstring), so it
        # needs its own fold: a per-stock Δamount books to *each* quantity the
        # stock carries (a CO2 pool's drop books to both CARBON and OXYGEN).
        # Without this, the per-step OXYGEN gate would falsely trip on every
        # photosynthesis step. A 1:1 stock folds ``delta · 1.0`` to its single
        # quantity — bit-identical to the pre-P2.1 path.
        for q, coeff in b.composition.items():
            bucket[q] = bucket.get(q, 0.0) + delta * coeff
    quantities = sorted(set(boundary) | set(stored), key=lambda q: q.name)
    return tuple(
        QuantityLedger(
            quantity=q,
            boundary_delta=boundary.get(q, 0.0),
            stored_delta=stored.get(q, 0.0),
            residual=boundary.get(q, 0.0) + stored.get(q, 0.0),
        )
        for q in quantities
    )


def assert_conserved(
    before: State,
    after: State,
    *,
    atol: float = BALANCE_ATOL,
    rtol: float = BALANCE_RTOL,
) -> None:
    """Raise ``ConservationError`` if any asserted quantity's mass is not conserved.

    The every-step engine gate (decision #7 / CLAUDE.md "conservation is asserted
    every step — a failure is a bug"). For each ``ASSERTED_QUANTITIES`` member (every
    ``Quantity`` except the exempt ``ENERGY``, #8), require
    ``abs(residual) <= atol + rtol * scale``, where ``scale`` is the transfer
    magnitude — ``max |per-stock Δ|`` for that quantity — matching
    ``flow.assert_flow_balanced``'s scale notion.

    What keeps this safe is **atol**, not the transfer magnitude. The Phase-0 demo
    *does* enter the watch-item's regime — ``Harvest`` fills the ``outside_c`` boundary
    reservoir to a large accumulated stock (~1e3) with shrinking deposits, so
    ``rtol·max|Δ|`` collapses toward the stored-rounding floor ``~eps·|amount|``. But
    ``tol >= atol`` (1e-9) sits ~3–4 orders above that floor, so the per-step ratio
    ``|residual|/tol`` is tiny: observed worst ≈ ``1.1e-4`` (a signed sum, so below the
    ceiling), analytic ceiling ``eps·Σ|amount| / atol`` ≈ ``2.4e-4`` —
    length-independent because conservation caps the total mass. An ``amount``-scaled
    basis (``Σ|amount|``) becomes the principled choice only once accumulated amounts
    approach ``atol/eps`` (~4.5e6), where ``eps·Σ|amount|`` overtakes ``atol`` — a
    flagged Phase-1 revisit, verified unnecessary for the Phase-0 demo (the step-11
    carry-forward gate in ``tests/test_biosphere_demo.py``).

    Thresholds the residual ``compute_ledger`` already computes (one residual
    computation; the ``ValueError`` key-set guard there also covers this entry point).
    """
    ledger = {ql.quantity: ql for ql in compute_ledger(before, after)}
    scale: dict[Quantity, float] = {}
    for sid, b in before.stocks.items():
        d = after.stocks[sid].amount - b.amount
        # Composition fold (P2.1), mirroring ``flow.assert_flow_balanced``'s scale:
        # the tolerance denominator for ``q`` is ``max |Δ · coeff_q|``. 1:1 folds
        # to ``abs(d)`` — unchanged.
        for q, coeff in b.composition.items():
            scale[q] = max(scale.get(q, 0.0), abs(d * coeff))
    # Sorted so the first reported failure is deterministic (a frozenset has no
    # defined iteration order) — cosmetic, but determinism everywhere is on-brand.
    for quantity in sorted(ASSERTED_QUANTITIES, key=lambda q: q.name):
        ql = ledger.get(quantity)
        if ql is None:
            continue  # quantity absent from the state — trivially conserved
        tol = atol + rtol * scale.get(quantity, 0.0)
        if abs(ql.residual) > tol:
            raise ConservationError(
                f"conservation violated for {quantity.name}: residual "
                f"{ql.residual!r} exceeds tolerance {tol!r} "
                f"(boundary_delta={ql.boundary_delta!r}, "
                f"stored_delta={ql.stored_delta!r})"
            )

File: rust/crates/simcore/src/conservation.rs
//! Conservation ledger + the every-step balance gate — the Rust port of
//! `simcore.conservation`.
//!
//! Over the augmented system (modeled stocks + boundary reservoirs), per asserted
//! quantity the total mass across all stocks is unchanged step-to-step, within
//! tolerance: `residual_q = Σ (after[s] − before[s]) ≈ 0`. This module reasons about
//! **state deltas, not flows** (it does not reuse `flow::assert_flow_balanced`), so it
//! has its own composition fold; it shares only the tolerance constants and the
//! `Conservation` error variant.
//!
//! Per-stock deltas are accumulated within each `StockKind` partition in **sorted
//! stock-id order** (the `BTreeMap` iteration order) — float sums are non-associative,
//! so this is what makes the ledger bit-identical under registration shuffle (#15).

use std::collections::HashMap;

use crate::error::SimError;
use crate::quantities::{Quantity, StockKind, ASSERTED_QUANTITIES, BALANCE_ATOL, BALANCE_RTOL};
use crate::state::State;

/// Per-quantity conservation accounting for one step. `residual == boundary_delta +
/// stored_delta` (≡ total-mass Δ) always holds by construction.
#[derive(Debug, Clone, PartialEq)]
pub struct QuantityLedger {
    pub quantity: Quantity,
    pub boundary_delta: f64,
    pub stored_delta: f64,
    pub residual: f64,
}

/// Per-quantity ledger for the step `before → after`, in name-sorted quantity order.
///
/// Returns [`SimError::Validation`] if `before`/`after` do not share the same stock-id
/// key set (Phase 0 never adds or removes stocks mid-run, so a mismatch is a bug).
pub fn compute_ledger(before: &State, after: &State) -> Result<Vec<QuantityLedger>, SimError> {
    if !same_key_set(before, after) {
        let before_only: Vec<&String> = before
            .stocks
            .keys()
            .filter(|k| !after.stocks.contains_key(*k))
            .collect();
        let after_only: Vec<&String> = after
            .stocks
            .keys()
            .filter(|k| !before.stocks.contains_key(*k))
            .collect();
        return Err(SimError::Validation(format!(
            "conservation ledger requires before/after to share the same stock ids; \
             Phase 0 never adds/removes stocks mid-run (before-only={before_only:?}, \
             after-only={after_only:?})"
        )));
    }
    let mut boundary: HashMap<Quantity, f64> = HashMap::new();
    let mut stored: HashMap<Quantity, f64> = HashMap::new();
    // BTreeMap iterates sorted by id — the canonical accumulation order (#15).
    for (sid, b) in &before.stocks {
        let delta = after.stocks[sid].amount - b.amount;
        let bucket = if b.kind == StockKind::Boundary {
            &mut boundary
        } else {
            &mut stored
        };
        for (q, coeff) in &b.composition {
            *bucket.entry(*q).or_insert(0.0) += delta * coeff;
        }
    }
    // Quantities present in either partition, in name-sorted order.
    let mut quantities: Vec<Quantity> = boundary
        .keys()
        .chain(stored.keys())
        .copied()
        .collect::<std::collections::BTreeSet<_>>()
        .into_iter()
        .collect();
    quantities.sort_by(|a, b| a.name().cmp(b.name()));
    Ok(quantities
        .into_iter()
        .map(|q| {
            let bd = boundary.get(&q).copied().unwrap_or(0.0);
            let sd = stored.get(&q).copied().unwrap_or(0.0);
            QuantityLedger {
                quantity: q,
                boundary_delta: bd,
                stored_delta: sd,
                residual: bd + sd,
            }
        })
        .collect())
}

fn same_key_set(before: &State, after: &State) -> bool {
    before.stocks.len() == after.stocks.len()
        && before.stocks.keys().all(|k| after.stocks.contains_key(k))
}

/// Raise [`SimError::Conservation`] if any asserted quantity's mass is not conserved
/// across the step (the every-step engine gate). `scale` is the transfer magnitude
/// (`max |per-stock Δ·coeff|`) for that quantity; `tol = atol + rtol*scale`.
pub fn assert_conserved(
    before: &State,
    after: &State,
    atol: f64,
    rtol: f64,
) -> Result<(), SimError> {
    let ledger: HashMap<Quantity, QuantityLedger> = compute_ledger(before, after)?
        .into_iter()
        .map(|ql| (ql.quantity, ql))
        .collect();
    let mut scale: HashMap<Quantity, f64> = HashMap::new();
    for (sid, b) in &before.stocks {
        let d = after.stocks[sid].amount - b.amount;
        for (q, coeff) in &b.composition {
            let acc = scale.entry(*q).or_insert(0.0);
            *acc = acc.max((d * coeff).abs());
        }
    }
    for quantity in ASSERTED_QUANTITIES {
        let ql = match ledger.get(&quantity) {
            Some(ql) => ql,
            None => continue, // quantity absent from the state — trivially conserved
        };
        let tol = atol + rtol * scale.get(&quantity).copied().unwrap_or(0.0);
        if ql.residual.abs() > tol {
            return Err(SimError::Conservation(format!(
                "conservation violated for {}: residual {:?} exceeds tolerance {tol:?} \
                 (boundary_delta={:?}, stored_delta={:?})",
                quantity.name(),
                ql.residual,
                ql.boundary_delta,
                ql.stored_delta
            )));
        }
    }
    Ok(())
}

/// The default-tolerance [`assert_conserved`] (using [`BALANCE_ATOL`] / [`BALANCE_RTOL`]).
pub fn assert_conserved_default(before: &State, after: &State) -> Result<(), SimError> {
    assert_conserved(before, after, BALANCE_ATOL, BALANCE_RTOL)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::state::Stock;
    use std::collections::BTreeMap;

    fn pool(id: &str, amount: f64) -> Stock {
        Stock::new(
            id.to_string(),
            "d".to_string(),
            Quantity::Carbon,
            "mol".to_string(),
            amount,
            StockKind::Pool,
            0.0,
            false,
            BTreeMap::new(),
        )
        .unwrap()
    }

    fn state(a: f64, b: f64) -> State {
        State::new(
            0,
            BTreeMap::from([("a".to_string(), pool("a", a)), ("b".to_string(), pool("b", b))]),
            0,
            BTreeMap::new(),
        )
        .unwrap()
    }

    #[test]
    fn balanced_transfer_conserves() {
        // a: 100 → 90, b: 0 → 10 — CARBON conserved.
        assert!(assert_conserved_default(&state(100.0, 0.0), &state(90.0, 10.0)).is_ok());
    }

    #[test]
    fn imbalanced_step_is_caught() {
        // a: 100 → 90, b: 0 → 5 — 5 mol CARBON vanished.
        let e = assert_conserved_default(&state(100.0, 0.0), &state(90.0, 5.0));
        assert!(matches!(e, Err(SimError::Conservation(_))));
    }

    fn co2(id: &str, amount: f64) -> Stock {
        // CO₂ = {Carbon:1, Oxygen:2} — the multi-quantity composition the ledger's
        // own `delta * coeff` fold must honour (cabin_gas / any biosphere CO₂ pool).
        Stock::new(
            id.to_string(),
            "d".to_string(),
            Quantity::Carbon,
            "mol".to_string(),
            amount,
            StockKind::Pool,
            0.0,
            false,
            BTreeMap::from([(Quantity::Carbon, 1.0), (Quantity::Oxygen, 2.0)]),
        )
        .unwrap()
    }

    #[test]
    fn compute_ledger_folds_composition_coeff() {
        // A CO₂ pool drops by 1 mol (nothing else changes). The OXYGEN delta must be
        // −2.0 (Δamount·coeff), not −1.0 — this is the exact `delta` vs `delta*coeff`
        // transcription a single-quantity scenario cannot catch (coeff = 1.0 hides it).
        let before = State::new(
            0,
            BTreeMap::from([("co2.a".to_string(), co2("co2.a", 10.0))]),
            0,
            BTreeMap::new(),
        )
        .unwrap();
        let after = State::new(
            0,
            BTreeMap::from([("co2.a".to_string(), co2("co2.a", 9.0))]),
            0,
            BTreeMap::new(),
        )
        .unwrap();
        let ledger: HashMap<Quantity, QuantityLedger> = compute_ledger(&before, &after)
            .unwrap()
            .into_iter()
            .map(|ql| (ql.quantity, ql))
            .collect();
        assert_eq!(ledger[&Quantity::Carbon].stored_delta, -1.0);
        assert_eq!(ledger[&Quantity::Oxygen].stored_delta, -2.0, "coeff-2 fold");
    }

    #[test]
    fn multi_quantity_balanced_transfer_conserves() {
        // co2.a → co2.b of 1 mol conserves both CARBON and OXYGEN (each side folds ·2
        // for oxygen), so the gate passes.
        let before = State::new(
            0,
            BTreeMap::from([
                ("co2.a".to_string(), co2("co2.a", 10.0)),
                ("co2.b".to_string(), co2("co2.b", 0.0)),
            ]),
            0,
            BTreeMap::new(),
        )
        .unwrap();
        let after = State::new(
            0,
            BTreeMap::from([
                ("co2.a".to_string(), co2("co2.a", 9.0)),
                ("co2.b".to_string(), co2("co2.b", 1.0)),
            ]),
            0,
            BTreeMap::new(),
        )
        .unwrap();
        assert!(assert_conserved_default(&before, &after).is_ok());
    }

    #[test]
    fn multi_quantity_imbalanced_step_is_caught() {
        // co2.a drops 1 mol but co2.b rises only 0.4 — both CARBON and OXYGEN break
        // (OXYGEN residual = −2 + 0.8 = −1.2), exercising the assert path over a
        // composition stock.
        let before = State::new(
            0,
            BTreeMap::from([
                ("co2.a".to_string(), co2("co2.a", 10.0)),
                ("co2.b".to_string(), co2("co2.b", 0.0)),
            ]),
            0,
            BTreeMap::new(),
        )
        .unwrap();
        let after = State::new(
            0,
            BTreeMap::from([
                ("co2.a".to_string(), co2("co2.a", 9.0)),
                ("co2.b".to_string(), co2("co2.b", 0.4)),
            ]),
            0,
            BTreeMap::new(),
        )
        .unwrap();
        assert!(matches!(
            assert_conserved_default(&before, &after),
            Err(SimError::Conservation(_))
        ));
    }

    #[test]
    fn keyset_mismatch_is_validation_error() {
        let before = state(1.0, 1.0);
        let after = State::new(
            0,
            BTreeMap::from([("a".to_string(), pool("a", 1.0))]),
            0,
            BTreeMap::new(),
        )
        .unwrap();
        assert!(matches!(
            compute_ledger(&before, &after),
            Err(SimError::Validation(_))
        ));
    }
}
